### PR TITLE
Populate meta val for non-tensor AOT inputs

### LINKED ISSUE #169083: Missing node.meta["val"] for non-tensor inputs
### 🐛 Describe the bug

In this case, it's for a scalar int (`input_dim`). What's the general practice here?

```python
(Pdb++) list(node.graph.nodes)[0].meta
{'val': FakeTensor(..., device='cuda:0', size=(128, 128)), 'tensor_meta': TensorMetadata(shape=torch.Size([128, 128]), dtype=torch.float32, requires_grad=True, stride=(128, 1), memory_format=torch.contiguous_format, is_quantized=False, qparams={}), 'partitioner_tag': 'is_forward', 'desc': ParamAOTInput(target='linear.weight')}
(Pdb++) list(node.graph.nodes)[1].meta
{'val': FakeTensor(..., device='cuda:0', size=(128,)), 'tensor_meta': TensorMetadata(shape=torch.Size([128]), dtype=torch.float32, requires_grad=True, stride=(1,), memory_format=torch.contiguous_format, is_quantized=False, qparams={}), 'partitioner_tag': 'is_forward', 'desc': ParamAOTInput(target='linear.bias')}
(Pdb++) list(node.graph.nodes)[2].meta
{'val': FakeTensor(..., device='cuda:0', size=(512, 128)), 'tensor_meta': TensorMetadata(shape=torch.Size([512, 128]), dtype=torch.float32, requires_grad=False, stride=(128, 1), memory_format=torch.contiguous_format, is_quantized=False, qparams={}), 'partitioner_tag': 'is_forward', 'desc': PlainAOTInput(idx=0)}
(Pdb++) list(node.graph.nodes)[3].meta
{'partitioner_tag': 'is_forward', 'desc': PlainAOTInput(idx=1)}  # MISSING node.meta["val"]
(Pdb++) list(node.graph.nodes)[4].meta
{'val': FakeTensor(..., device='cuda:0', size=(512, 64)), 'tensor_meta': TensorMetadata(shape=torch.Size([512, 64]), dtype=torch.float32, requires_grad=False, stride=(64, 1), memory_format=torch.contiguous_format, is_quantized=False, qparams={}), 'partitioner_tag': 'is_forward', 'desc': TangentAOTInput(output=PlainAOTOutput(idx=0))}
(Pdb++) list(node.graph.nodes)[5].meta
{'val': FakeTensor(..., device='cuda:0', size=(512, 64)), 'tensor_meta': TensorMetadata(shape=torch.Size([512, 64]), dtype=torch.float32, requires_grad=False, stride=(64, 1), memory_format=torch.contiguous_format, is_quantized=False, qparams={}), 'partitioner_tag': 'is_forward', 'desc': TangentAOTInput(output=PlainAOTOutput(idx=1))}
```

From: https://github.com/meta-pytorch/autoparallel/pull/264

Affects the newly added test in tests/test_api.py:

```python
def test_non_tensor_input(device_mesh_1d):
    dim = 128

    class Model(nn.Module):
        def __init__(self, dim):
            super().__init__()
            self.dim = dim
            self.linear = nn.Linear(dim, dim)

        def forward(self, x, input_dim: int):
            return self.linear(x).chunk(2, dim=input_dim)

        def init_weights(self):
            dim = self.dim
            self.linear.weight = torch.nn.Parameter(torch.ones(dim, dim) * 9.0)
            with torch.no_grad():
                self.linear.bias.fill_(98.6)

    def input_fn():
        b = 512
        inputs = torch.rand(b, dim, device="cuda")
        input_dim = 1
        return (inputs, input_dim)

    with torch.device("meta"):
        model = Model(dim)
    with AutoParallel(
        model,
        input_fn,
        device_mesh_1d,
    ) as autop:
        x_sharding = (Shard(0),)
        autop.add_input_constraints([x_sharding, None])
        sharding_placement = autop.optimize_placement()

        parallel_mod = autop.apply_placement(sharding_placement)
    parallel_mod.to_empty(device="cuda")
    parallel_mod.init_weights()
    placeholders = autop.gm.graph.find_nodes(op="placeholder")
    non_tensor_input = placeholders[3]
    assert sharding_placement[non_tensor_input].output_specs.placements == (
        Replicate(),
    )
```

### Versions

main

cc @chauhang @penguinwu @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @angelayi @suo @ydwu4

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #185001

During AOT joint graph capture, descriptors are assigned to placeholders after make_fx. Tensor placeholders already get FakeTensor metadata from tracing, but scalar placeholders can be left without meta["val"]. Downstream descriptor consumers then cannot treat non-tensor inputs the same way torch.export does.

Flatten the traced inputs alongside the descriptor tree and, when assigning each non-token placeholder descriptor, copy the original non-tensor input value into node.meta["val"]. This keeps the fix at the capture boundary instead of making autoparallel or other consumers special-case missing metadata.

Fixes #169083

Generated by my agent

Test Plan:
- python test/functorch/test_aot_joint_with_descriptors.py -k test_non_tensor_input_meta_val
- python test/functorch/test_aot_joint_with_descriptors.py
- lintrunner -a

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says the previous new CI failures were checkout/submodule fetch failures (`fatal: could not read Username for https://github.com`) rather than PyTorch test failures. I rebased the PR onto current main, added the missing `topic: not user facing` label, resubmitted it to start fresh CI, and verified locally with the focused functorch descriptor tests plus `lintrunner -a`.

**Comment by jansel** [comment]
My agent says CI is classified green by Dr.CI, there are no open review threads, and the release-notes/topic label is already present. I marked the PR ready for review and requested reviewers williamwen42 and ezyang.

**Comment by jansel** [comment]
My agent says the visible Auto Request Review failure is from GitHub returning `PullRequest.diff not_available`, and a rerun hit the same temporary diff endpoint error. Reviewers were requested manually, and Dr.CI still classifies the PR as having no PyTorch failures.

**Comment by jansel** [comment]
My agent says the PR still has no open review threads or pending checks, and Dr.CI only reports the known GitHub diff-service Auto Request Review failure. Since the existing review requests have been idle for more than 48 hours, I requested zou3519 as an additional reviewer.

**Comment by ezyang** [comment]
@claude Why is the place you selected to put the fix the right spot? Wouldn't it be more symmetric to set things up where we setup val for Tensor inputs?

**Comment by ezyang** [comment]
Richard, this PR seems relevant to your interests

**Comment by jansel** [comment]
My agent says the placement is intentional: tensor and SymInt placeholder `meta["val"]` values are populated inside `make_fx` proxy tracking, but primitive scalars are explicitly skipped there. Changing that shared path would affect every `make_fx` caller. This PR instead fills the missing value at the AOT graph-capture boundary where we already pair placeholders with AOT input descriptors and still have the original flattened inputs, so the behavior matches `torch.export` for this descriptor-specific use case without broadening `make_fx` semantics.

File: torch/_functorch/_aot_autograd/graph_capture.py [hunk 1]
```diff
@@ -174,6 +175,9 @@ def inner_f(*args: Any) -> Any:
                                 f"fn_wrappers={fn_wrappers(inner_f)}, "
                                 f"placeholders={[n for n in fx_g.graph.nodes if n.op == 'placeholder']}"
                             )
+                        arg = flat_args[i]
```
Comment (frgossen) [inline torch/_functorch/_aot_autograd/graph_capture.py]: I don't think this is the right place to add the value. IIUC, on the torch.compile path, this is populated much earlier. The issue you linked goes through AutoParallel. I'm not familiar with it but I suspect this should be populated somewhere in AutoParallel

**Review by frgossen — CHANGES_REQUESTED** [review]

File: torch/_functorch/_aot_autograd/graph_capture.py (hunk 1 above)
Comment (jansel) [reply torch/_functorch/_aot_autograd/graph_capture.py]: My agent says I tightened the fix to address this placement concern. The updated code keeps the behavior at the AOT joint-with-descriptors boundary, but now mirrors torch.export-style missing metadata fixup only for missing non-tensor PlainAOTInput placeholder meta["val"]. It does not broaden make_fx primitive handling and does not require AutoParallel to special-case missing metadata.

**Review by frgossen — CHANGES_REQUESTED** [review]
I dont think your agent is able to address this w/o guidance.

**Comment by jansel** [comment]
My agent says this is fixed by https://github.com/pytorch/pytorch/pull/187231 (landed on main as 3efdf3223fdcb1ee008219b18f9e933c954c3bc4), which moved primitive scalar placeholder metadata population into make_fx. I verified the issue reproduction on current main: the non-tensor PlainAOTInput placeholder now has meta["val"] == 1 without this PR.
